go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// processAliasedTable produces a PlanBuilder subtree and SymbolTable for the given AliasedTableExpr.
+func processAliasedTable(tableExpr *sqlparser.AliasedTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch expr := tableExpr.Expr.(type) {
+	case *sqlparser.TableName:
+		route, table, err := getTablePlan(expr, schema)
+		if err != nil {
+			return nil, nil, err
+		}
+		planBuilder := &RouteBuilder{
+			From:  tableExpr,
+			Route: route,
+		}
+		alias := expr.Name
+		if tableExpr.As != "" {
+			alias = tableExpr.As
+		}
+		symbols := NewSymbolTable(alias, table, planBuilder)
+		return planBuilder, symbols, nil
+	case *sqlparser.Subquery:
+		return nil, nil, errors.New("no subqueries")
+	}
+	panic("unreachable")
+}
+
+// getTablePlan produces the initial Route for the specified TableName. It also returns
+// the associated vschema info (*Table) so that it can be used to create the symbol table
+// entry.
+func getTablePlan(tableName *sqlparser.TableName, schema *Schema) (*Route, *Table, error) {
+	if tableName.Qualifier != "" {
+		return nil, nil, errors.New("tablename qualifier not allowed")
+	}
+	table, reason := schema.FindTable(string(tableName.Name))
+	if reason != "" {
+		return nil, nil, errors.New(reason)
+	}
+	if table.Keyspace.Sharded {
+		return &Route{
+			PlanID:   SelectScatter,
+			Keyspace: table.Keyspace,
+		}, table, nil
+	}
+	return &Route{
+		PlanID:   SelectUnsharded,
+		Keyspace: table.Keyspace,
+	}, table, nil
+}
+
+// processJoin produces a PlanBuilder subtree and SymbolTable for the given Join. If the left and
+// right node can be merged into one chunk, then it's a RouteBuilder. Otherwise, it's a JoinBuilder.
+func processJoin(join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch join.Join {
+	case sqlparser.JoinStr, sqlparser.StraightJoinStr, sqlparser.LeftJoinStr:
+	default:
+		return nil, nil, errors.New("unsupported join")
+	}
+	lplanBuilder, lsymbols, err := processTableExpr(join.LeftExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	rplanBuilder, rsymbols, err := processTableExpr(join.RightExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	switch lplanBuilder := lplanBuilder.(type) {
+	case *JoinBuilder:
+		return makeChunk(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+	case *RouteBuilder:
+		switch rplanBuilder := rplanBuilder.(type) {
+		case *JoinBuilder:
+			return makeChunk(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+		case *RouteBuilder:
+			return joinRoutes(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join, schema)
+		}
+	}
+	panic("unreachable")
+}
+
+// makeChink creates a new chunk for the join operation. This function is called when
+// two chunks cannot be merged into one.
+func makeChunk(lplanBuilder PlanBuilder, lsymbols *SymbolTable, rplanBuilder PlanBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
```

Comment:
The function name seems misleading to me. You’re not really creating a new chunk, but just joining two existing chunks and merging their symbol tables. Maybe call it makeJoinBuilder instead?